alpha/veneer/composite/builder.go

```diff
@@ -0,0 +1,287 @@
+package composite
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path"
+
+	"github.com/operator-framework/operator-registry/alpha/declcfg"
+)
+
+const (
+	BasicVeneerBuilderSchema  = "olm.veneer.basic"
+	SemverVeneerBuilderSchema = "olm.veneer.semver"
+	RawVeneerBuilderSchema    = "olm.veneer.raw"
+	CustomVeneerBuilderSchema = "olm.veneer.custom"
+)
+
+type ContainerConfig struct {
+	ContainerTool string
+	BaseImage     string
+	WorkingDir    string
+}
+
+type BuilderConfig struct {
+	ContainerCfg ContainerConfig
+	OutputType   string
+}
+
+type Builder interface {
+	Build(dir string, vd VeneerDefinition) error
+	Validate(dir string) error
+}
+
+type BasicBuilder struct {
+	builderCfg BuilderConfig
+}
+
+var _ Builder = &BasicBuilder{}
+
+func NewBasicBuilder(builderCfg BuilderConfig) *BasicBuilder {
+	return &BasicBuilder{
+		builderCfg: builderCfg,
+	}
+}
+
+func (bb *BasicBuilder) Build(dir string, vd VeneerDefinition) error {
+	// Parse out the basic veneer configuration
+	basicConfig := &BasicVeneerConfig{}
+	err := json.Unmarshal(vd.Config, basicConfig)
+	if err != nil {
+		return fmt.Errorf("unmarshalling basic veneer config: %w", err)
+	}
+	// TODO: should we ensure that the schema matches and error if it does not?
+
+	// get the current working directory
+	wd, err := os.Getwd()
+	if err != nil {
+		return fmt.Errorf("getting current working directory: %w", err)
+	}
+
+	// build the container command
+	containerCmd := exec.Command(bb.builderCfg.ContainerCfg.ContainerTool,
+		"run",
+		"--rm",
+		"-v",
+		fmt.Sprintf("%s:%s", wd, bb.builderCfg.ContainerCfg.WorkingDir),
+		bb.builderCfg.ContainerCfg.BaseImage,
+		"alpha",
+		"render-veneer",
+		"basic",
+		path.Join(bb.builderCfg.ContainerCfg.WorkingDir, basicConfig.Input))
+
+	return build(containerCmd, path.Join(dir, basicConfig.Output), bb.builderCfg.OutputType)
```

Comment:
Do we panic if the bb.* fields are not present in the file?  Do we need to verify that we have valid values (at a minimum, not empty-string) before we use them?

Reply:
It shouldn't panic because they aren't pointers but running the command would likely fail if it was an empty string. I could probably update the structs to have some kind of json/yaml parsing tags that requires fields to be present or else it errors when unmarshalling.

If the command fails for any reason it will return an error here: https://github.com/operator-framework/operator-registry/pull/1054/files#diff-932fc031fb99963dd1cedfe8ae41f2c1531dcfedba6e36f2c396e28e7403ba21R263-R266

Reply:
So I took a brief look into this and I was correct that running the command ends up failing. I also did some research on Go struct tags and it doesn't look like there is an easy way to specify that a field is required without adding additional dependencies and extra struct tags.

What I am thinking to do is just add some checks that will fail fast if the WorkingDir and BaseImage fields are empty strings prior to the creation of a builder in this loop:
https://github.com/operator-framework/operator-registry/pull/1054/files#diff-f0ad1e448a4ece45a9aa1e9e28bafc35d5c88ffd8517dadbe2a679baa02ba89eR58-R78

Reply:
fail fast is good as a first pass for sure. but something to keep in mind from a UX perspective. if someone has multiple errors in their input file, its nice to tell them about all of them rather than just the first one you encounter.

Reply:
Added validations and instead of failing fast I created the validations in such a way that will check all of the fields before failing and append them to an error message.

All this and the respective test cases were added in https://github.com/operator-framework/operator-registry/pull/1054/commits/f33598f9912c7e2cfa9d55ec79d54c0fd174e1b3